see how a decimal number is represented in binary

In [1]:
'{0:b}'.format(194) #2^7 + 2^6 + 2^1

'11000010'

perform bitwise XOR

In [6]:
10^200 

194

compares the bits and gives a 1 as output if the output matches with XOR:

i.e. 00001010 and 11001000 are compared, i.e. 10 and 200 and compared (..0+8+4*0+2*1+0=10 and 128+64+0+0+8+0..=200) and 194 is returned, i.e. 11000010 = 128+64+2

represent a decimal number in binary:

In [13]:
'{0:b}'.format(194)

'11000010'

These two are the same: aas the first n cubed integeres are equal to the sum of the n squared

In [9]:
1**3+2**3+3**3

36

In [14]:
(1+2+3)**2

36

variables: have an assigned value and a type, the increment operator adds to the variable the consequent value

In [15]:
a=5
a+=5 # equal to a = a + 5
a

10

convert a string containing numbers from the actual string to an integer

In [23]:
my_str = '42'
my_int = int(my_str)
print(my_int, type(my_int))

42 <class 'int'>


or convert the string to an integer

In [24]:
str(my_int)
print(my_int, type(my_int))

42 <class 'int'>


In [25]:
type(my_int) ## this didnt worked

int

converting a float to an integer just gives us the floor and doesnt rounds the result

In [27]:
int(2.999)

2

In [29]:
#watch out for this too
print('4' + '2')
print(int('4') + int('2'))

42
6


<i>Never use the == operator with floats.</i>

In [ ]:
!= #not equal

Unicode based strings

In [32]:
ord('a') #assess number associated with character
ord('λ')

# relational operators compare the values that the ord() function is returning

'a'<'λ' #true

True

Python allow chaining of relational operators.

In [37]:
4 < 6 < 6.1 < 9.3

#important not to do the following, even though it is legal
4 < 6.1 > 5

True

<i>Do not mix the directions of chained relational operators.</i>

<b>identity operators</b>

In [46]:
a=3.0
b=3.0

a is b, a is not b

(False, True)

happens to store the two values in two different places in memory, so if one wants to compare the floats....

In [47]:
a = 5.6
b = a

a == b, a is b

(True, True)

for integers, Python employs integer caching for all integers between -5 and 256... so they are stored in the same place

In [50]:
a = 257
b = 257
c = 256
d = 256

a is b, c is d

(False, True)

same story for strings, it depends on which string to be stored either in the same place or not.


<b>precedence of operators </b>

1  **

2  *, /, //, %

3  +, -

4  <, >, <=, >=

5  ==, !=

6  =, +=, -=, *=, /=, **=, %=, //=

7  is, is not

8  and, or, not

mebership operators, come later:
in
not in

not in bootcamp: bitwise operators and operators on sets